gdal/alg/gdalrasterize.cpp

```diff
@@ -670,90 +703,260 @@ CPLErr GDALRasterizeGeometries( GDALDatasetH hDS,
+                                        padfGeomBurnValue + iShape*nBandCount,
+                                        eBurnValueSource, eMergeAlg,
+                                        pfnTransformer, pTransformArg );
+            }
+    
+            eErr =
+                poDS->RasterIO( GF_Write, 0, iY,
+                                poDS->GetRasterXSize(), nThisYChunkSize,
+                                pabyChunkBuf,
+                                poDS->GetRasterXSize(), nThisYChunkSize,
+                                eType, nBandCount, panBandList, 0, 0, 0, NULL);
+    
+            if( !pfnProgress((iY + nThisYChunkSize) /
+                             static_cast<double>(poDS->GetRasterYSize()),
+                             "", pProgressArg ) )
+            {
+                CPLError( CE_Failure, CPLE_UserInterrupt, "User terminated" );
+                eErr = CE_Failure;
+            }
+        }
+    }
+/* -------------------------------------------------------------------- */
+/*      The new algorithm                                               */
+/*      Optimized to minimize the vector computation                    */
+/*      (optimal on a large number of vectors & tiled raster)           */
+/* -------------------------------------------------------------------- */
+    else
+    {
+/* -------------------------------------------------------------------- */
+/*      Establish a chunksize to operate on.  Its size is defined by    */
+/*      the block size of the output file.                              */
+/* -------------------------------------------------------------------- */
+        const int nXBlocks = (poBand->GetXSize() + nXBlockSize - 1) / nXBlockSize;
+        const int nYBlocks = (poBand->GetYSize() + nYBlockSize - 1) / nYBlockSize;
 
-    pfnProgress( 0.0, NULL, pProgressArg );
 
-    for( int iY = 0;
-         iY < poDS->GetRasterYSize() && eErr == CE_None;
-         iY += nYChunkSize )
-    {
-        int nThisYChunkSize = nYChunkSize;
-        if( nThisYChunkSize + iY > poDS->GetRasterYSize() )
-            nThisYChunkSize = poDS->GetRasterYSize() - iY;
-
-        eErr =
-            poDS->RasterIO(GF_Read,
-                           0, iY, poDS->GetRasterXSize(), nThisYChunkSize,
-                           pabyChunkBuf,
-                           poDS->GetRasterXSize(), nThisYChunkSize,
-                           eType, nBandCount, panBandList,
-                           0, 0, 0, NULL);
-        if( eErr != CE_None )
-            break;
+        const GDALDataType eType =
+            poBand->GetRasterDataType() == GDT_Byte ? GDT_Byte : GDT_Float64;
+
+        const int nPixelSize = nBandCount * GDALGetDataTypeSizeBytes(eType);
+
+        // TODO use GDAL_CACHEMAX for the buffer size ?
+        // rem: optimized for square blocks
+        const int nbMaxBlocks = 500000000 / nPixelSize / nYBlockSize / nXBlockSize;
```

Comment:
use atoi(CSLFetchNameValue(papszOptions, "BUFFERSIZE", "500000000")) instead of the hardcoded value